File: examples/gods_of_the_arena/neural_package.nim
## GotA neural package (gota-neural-basic/1): a ZIP of exactly manifest.json,
## policy.bas and model.bin. Parsing is strict: any unknown key, missing
## file, hash mismatch or contract mismatch rejects the package. The Python
## staging validator (coworld/gota/runtime/neural_package.py) mirrors these
## rules. See neural_basic.md.

import
  std/[json, strutils],
  crunchy, zippy,
  neural_actor, neural_contract

const
  PackageSchema* = "gota-neural-basic/1"
  PackageFiles* = ["manifest.json", "policy.bas", "model.bin"]
  MaxPackageBytes* = 16 * 1024 * 1024
  MaxPolicyBytes* = 256 * 1024
  ZipMagic* = "PK\x03\x04"

type
  DecoderMode* = enum ArgmaxDecoder, SampleDecoder
  NeuralPackage* = object
    decisionPeriod*: int32
    policy*: string
    model*: string
    actor*: Actor
    goals*: array[2, array[GoalSize, float32]] ## red, blue
    decoder*: DecoderMode
    temperature*: float32
    manifest*: string

proc isPackage*(bytes: string): bool = bytes.startsWith(ZipMagic)

proc sha256Hex*(data: string): string =
  for value in sha256(cast[pointer](data.cstring), data.len):
    result.add value.toHex(2).toLowerAscii()

proc u16(s: string, p: int): int =
  if p + 2 > s.len: raise newException(ValueError, "truncated zip")
  ord(s[p]) or (ord(s[p+1]) shl 8)

proc u32(s: string, p: int): int =
  if p + 4 > s.len: raise newException(ValueError, "truncated zip")
  ord(s[p]) or (ord(s[p+1]) shl 8) or (ord(s[p+2]) shl 16) or (ord(s[p+3]) shl 24)

proc readZip*(bytes: string): seq[(string, string)] =
  ## Reads every entry of a small, unencrypted, single-disk ZIP (stored or
  ## deflate) through its central directory.
  if bytes.len > MaxPackageBytes:
    raise newException(ValueError, "package exceeds 16 MiB")
  var eocd = -1
  for p in countdown(bytes.len - 22, max(0, bytes.len - 22 - 65535)):
    if bytes[p] == 'P' and bytes[p+1] == 'K' and bytes[p+2] == '\x05' and bytes[p+3] == '\x06':
      eocd = p
      break
  if eocd < 0: raise newException(ValueError, "zip end record not found")
  let
    entries = u16(bytes, eocd + 10)
    cdOffset = u32(bytes, eocd + 16)
  if entries > 16: raise newException(ValueError, "too many zip entries")
  var p = cdOffset
  for _ in 0 ..< entries:
    if u32(bytes, p) != 0x02014b50: raise newException(ValueError, "bad zip directory")
    let
      flags = u16(bytes, p + 8)
      methodId = u16(bytes, p + 10)
      compressed = u32(bytes, p + 20)
      size = u32(bytes, p + 24)
      nameLen = u16(bytes, p + 28)
      extraLen = u16(bytes, p + 30)
      commentLen = u16(bytes, p + 32)
      local = u32(bytes, p + 42)
    if p + 46 + nameLen > bytes.len: raise newException(ValueError, "truncated zip")
    let name = bytes[p + 46 ..< p + 46 + nameLen]
    if (flags and 1) != 0: raise newException(ValueError, "encrypted zip entry")
    if u32(bytes, local) != 0x04034b50: raise newException(ValueError, "bad zip entry")
    let
      dataStart = local + 30 + u16(bytes, local + 26) + u16(bytes, local + 28)
    if dataStart + compressed > bytes.len: raise newException(ValueError, "truncated zip data")
    let raw = bytes[dataStart ..< dataStart + compressed]
    var data: string
    case methodId
    of 0: data = raw
    of 8: data = uncompress(raw, dfDeflate)
    else: raise newException(ValueError, "unsupported zip compression " & $methodId)
    if data.len != size: raise newException(ValueError, "zip size mismatch for " & name)
    result.add((name, data))
    p += 46 + nameLen + extraLen + commentLen

proc requireKeys(node: JsonNode, allowed: openArray[string], where: string) =
  if node.kind != JObject:
    raise newException(ValueError, where & " must be an object")
  for key in node.keys:
    if key notin allowed:
      raise newException(ValueError, "unknown manifest key " & where & "." & key)

proc goalVector(node: JsonNode, where: string): array[GoalSize, float32] =
  if node.kind != JArray or node.len != GoalSize:
    raise newException(ValueError, where & " must be 16 numbers")
  for i in 0 ..< GoalSize:
    if node[i].kind notin {JInt, JFloat}:
      raise newException(ValueError, where & " must be 16 numbers")
    let v = node[i].getFloat
    if v < -1 or v > 1:
      raise newException(ValueError, where & " values must be in [-1, 1]")
    result[i] = float32(v)
  if result[GoalSize - 1] != 0:
    raise newException(ValueError, where & " w_reserved must be 0")

proc defaultGoal*(): array[GoalSize, float32] =
  result[0] = 1

proc parsePackage*(bytes: string): NeuralPackage =
  ## Validates a package completely; raises ValueError with the reason.
  let entries = readZip(bytes)
  if entries.len != 3:
    raise newException(ValueError, "package must contain exactly manifest.json, policy.bas and model.bin")
  var files: array[3, string]
  var seen: array[3, bool]
  for (name, data) in entries:
    let index = PackageFiles.find(name)
    if index < 0 or seen[index]:
      raise newException(ValueError, "unexpected package entry " & name)
    seen[index] = true
    files[index] = data
  let manifest = parseJson(files[0])
  manifest.requireKeys(["schema", "observation_contract", "action_contract",
    "decision_period", "files", "model", "goal", "decoder"], "manifest")
  for key in ["schema", "observation_contract", "action_contract",
      "decision_period", "files", "model"]:
    if not manifest.hasKey(key):
      raise newException(ValueError, "manifest is missing " & key)
  if manifest["schema"].getStr != PackageSchema:
    raise newException(ValueError, "manifest schema must be " & PackageSchema)
  if manifest["observation_contract"].getStr != ObservationContractHash:
    raise newException(ValueError, "observation contract mismatch")
  if manifest["action_contract"].getStr != ActionContractHash:
    raise newException(ValueError, "action contract mismatch")
  let period = manifest["decision_period"]
  if period.kind != JInt or period.getInt notin 1..24:
    raise newException(ValueError, "decision_period must be an integer 1..24")
  result.decisionPeriod = int32(period.getInt)
  let fileHashes = manifest["files"]
  fileHashes.requireKeys(["policy.bas", "model.bin"], "files")
  if not fileHashes.hasKey("policy.bas") or not fileHashes.hasKey("model.bin"):
    raise newException(ValueError, "files must list policy.bas and model.bin")
  if fileHashes["policy.bas"].getStr != sha256Hex(files[1]):
    raise newException(ValueError, "policy.bas sha256 mismatch")
  if fileHashes["model.bin"].getStr != sha256Hex(files[2]):
    raise newException(ValueError, "model.bin sha256 mismatch")
  if files[1].len > MaxPolicyBytes:
    raise newException(ValueError, "policy.bas exceeds 256 KiB")
  let model = manifest["model"]
  model.requireKeys(["format", "inputs", "hidden", "heads"], "model")
  for key in ["format", "inputs", "hidden", "heads"]:
    if not model.hasKey(key):
      raise newException(ValueError, "model is missing " & key)
  if model["format"].getStr != "GOTANET1":
    raise newException(ValueError, "model.format must be GOTANET1")
  let actor = loadActor(files[2])
  if actor.observationContract != ObservationContractHash or
      actor.actionContract != ActionContractHash:
    raise newException(ValueError, "model.bin contract hashes do not match the manifest")
  if model["inputs"].getInt != actor.inputSize or actor.inputSize != ObservationSize:
    raise newException(ValueError, "model.inputs must be " & $ObservationSize)
  if model["hidden"].getInt != actor.hiddenSize:
    raise newException(ValueError, "model.hidden does not match model.bin")
  var heads: seq[int]
  for h in model["heads"]:
    heads.add h.getInt
  if heads != @HeadSizes or actor.headSizes != @HeadSizes:
    raise newException(ValueError, "model.heads must be [8,25,49,4,6]")
  result.actor = actor
  result.goals = [defaultGoal(), defaultGoal()]
  if manifest.hasKey("goal"):
    let goal = manifest["goal"]
    goal.requireKeys(["red", "blue"], "goal")
    if not goal.hasKey("red") or not goal.hasKey("blue"):
      raise newException(ValueError, "goal needs red and blue")
    result.goals[0] = goalVector(goal["red"], "goal.red")
    result.goals[1] = goalVector(goal["blue"], "goal.blue")
  result.decoder = ArgmaxDecoder
  result.temperature = 1
  if manifest.hasKey("decoder"):
    let decoder = manifest["decoder"]
    decoder.requireKeys(["mode", "temperature"], "decoder")
    let mode = if decoder.hasKey("mode"): decoder["mode"].getStr else: "argmax"
    case mode
    of "argmax":
      if decoder.hasKey("temperature"):
        raise newException(ValueError, "decoder.temperature needs mode sample")
    of "sample":
      result.decoder = SampleDecoder
      if decoder.hasKey("temperature"):
        let t = decoder["temperature"].getFloat
        if t < 0.01 or t > 10:
          raise newException(ValueError, "decoder.temperature must be 0.01..10")
        result.temperature = float32(t)
    else:
      raise newException(ValueError, "decoder.mode must be argmax or sample")
  result.policy = files[1]
  result.model = files[2]
  result.manifest = files[0]

File: coworld/gota/runtime/neural_package.py
"""Staging validator and builder for GotA neural packages (gota-neural-basic/1).

A package is a ZIP holding exactly manifest.json, policy.bas and model.bin. This module
mirrors examples/gods_of_the_arena/neural_package.nim rule for rule, so a submission the
staging step accepts is one the game accepts (and the reverse). Plain .bas submissions are not
packages and are untouched.

  python3 neural_package.py validate PACKAGE.zip [--obs-hash H --action-hash H]
  python3 neural_package.py build OUT.zip --policy policy.bas --model model.bin
      [--period 4] [--decoder argmax|sample --temperature T] [--goal-red 16 floats] [--goal-blue ...]
"""
import argparse, hashlib, io, json, math, struct, sys, zipfile

SCHEMA = "gota-neural-basic/1"
FILES = ("manifest.json", "policy.bas", "model.bin")
MAX_PACKAGE = 16 * 1024 * 1024
MAX_POLICY = 256 * 1024
HEADS = [8, 25, 49, 4, 6]
OBS_SIZE = 1407
WIDTHS = (64, 128, 256)
MAGIC = b"GOTANET1"
OP_BUDGET = 4_000_000
MAX_PARAMS = 2_000_000
# Contract v1 hashes (examples/gods_of_the_arena/neural_contract.nim; gota_*_contract_hash).
OBS_HASH = "ae4046e83cc02e861f9c8cc32550c6cc4d6f9c161c9225a9b34a314d310ea991"
ACTION_HASH = "ecc7d53c11a9db0912467c66ecb3e65b60b3e71ef14dad1442ba3b4f6ac14697"
TOP_KEYS = {"schema", "observation_contract", "action_contract", "decision_period", "files", "model", "goal", "decoder"}
REQUIRED = ("schema", "observation_contract", "action_contract", "decision_period", "files", "model")


class PackageError(ValueError):
    pass


def is_package(data: bytes) -> bool:
    return data[:4] == b"PK\x03\x04"


def _keys(node, allowed, where):
    if not isinstance(node, dict):
        raise PackageError(f"{where} must be an object")
    for key in node:
        if key not in allowed:
            raise PackageError(f"unknown manifest key {where}.{key}")


def _goal(node, where):
    if not isinstance(node, list) or len(node) != 16 or not all(
            isinstance(v, (int, float)) and not isinstance(v, bool) for v in node):
        raise PackageError(f"{where} must be 16 numbers")
    if any(v < -1 or v > 1 for v in node):
        raise PackageError(f"{where} values must be in [-1, 1]")
    if node[15] != 0:
        raise PackageError(f"{where} w_reserved must be 0")


def check_model(data: bytes, obs_hash=OBS_HASH, action_hash=ACTION_HASH):
    """Validates a GOTANET1 model.bin; returns (hidden, operations)."""
    if data[:8] != MAGIC:
        raise PackageError("invalid neural actor magic (want GOTANET1)")
    if len(data) < 32:
        raise PackageError("truncated neural actor")
    version, inputs, hidden, outputs, heads, params = struct.unpack_from("<6I", data, 8)
    if version != 1 or not 1 <= inputs <= 4096 or hidden not in WIDTHS or not 2 <= outputs <= 1024 or not 1 <= heads <= 32:
        raise PackageError("unsupported neural actor dimensions/version")
    expected = inputs * hidden + 3 * hidden * hidden + outputs * hidden
    if params != expected or expected > MAX_PARAMS or len(data) != 32 + 128 + 4 * heads + 4 * expected:
        raise PackageError("invalid neural actor length/parameter count")
    ohash, ahash = data[32:96].decode("ascii", "replace"), data[96:160].decode("ascii", "replace")
    for h in (ohash, ahash):
        if any(c not in "0123456789abcdef" for c in h):
            raise PackageError("invalid neural contract hash")
    sizes = list(struct.unpack_from(f"<{heads}I", data, 160))
    if any(not 2 <= s <= 1024 for s in sizes) or sum(sizes) != outputs:
        raise PackageError("head/output mismatch")
    weights = memoryview(data)[160 + 4 * heads:].cast("f")
    if not all(math.isfinite(w) for w in weights):
        raise PackageError("nonfinite neural weight")
    ops = 2 * expected + 32 * hidden
    if ops > OP_BUDGET:
        raise PackageError(f"neural actor needs {ops} operations per inference, over the {OP_BUDGET} budget")
    if ohash != obs_hash or ahash != action_hash:
        raise PackageError("model.bin contract hashes do not match the manifest")
    if inputs != OBS_SIZE or sizes != HEADS:
        raise PackageError("model inputs/heads do not match contract v1")
    return hidden, ops


def validate(data: bytes, obs_hash=OBS_HASH, action_hash=ACTION_HASH) -> dict:
    """Raises PackageError with the reason; returns the manifest on success."""
    if len(data) > MAX_PACKAGE:
        raise PackageError("package exceeds 16 MiB")
    try:
        z = zipfile.ZipFile(io.BytesIO(data))
    except zipfile.BadZipFile as e:
        raise PackageError(f"not a zip: {e}")
    infos = z.infolist()
    names = [i.filename for i in infos]
    if len(infos) != 3 or sorted(names) != sorted(FILES):
        raise PackageError("package must contain exactly manifest.json, policy.bas and model.bin")
    for i in infos:
        if i.flag_bits & 1:
            raise PackageError("encrypted zip entry")
        if i.compress_type not in (zipfile.ZIP_STORED, zipfile.ZIP_DEFLATED):
            raise PackageError(f"unsupported zip compression {i.compress_type}")
    files = {n: z.read(n) for n in FILES}
    try:
        m = json.loads(files["manifest.json"])
    except ValueError as e:
        raise PackageError(f"manifest.json is not JSON: {e}")
    _keys(m, TOP_KEYS, "manifest")
    for key in REQUIRED:
        if key not in m:
            raise PackageError(f"manifest is missing {key}")
    if m["schema"] != SCHEMA:
        raise PackageError(f"manifest schema must be {SCHEMA}")
    if m["observation_contract"] != obs_hash:
        raise PackageError("observation contract mismatch")
    if m["action_contract"] != action_hash:
        raise PackageError("action contract mismatch")
    p = m["decision_period"]
    if not isinstance(p, int) or isinstance(p, bool) or not 1 <= p <= 24:
        raise PackageError("decision_period must be an integer 1..24")
    _keys(m["files"], {"policy.bas", "model.bin"}, "files")
    if set(m["files"]) != {"policy.bas", "model.bin"}:
        raise PackageError("files must list policy.bas and model.bin")
    for n in ("policy.bas", "model.bin"):
        if m["files"][n] != hashlib.sha256(files[n]).hexdigest():
            raise PackageError(f"{n} sha256 mismatch")
    if len(files["policy.bas"]) > MAX_POLICY:
        raise PackageError("policy.bas exceeds 256 KiB")
    model = m["model"]
    _keys(model, {"format", "inputs", "hidden", "heads"}, "model")
    for key in ("format", "inputs", "hidden", "heads"):
        if key not in model:
            raise PackageError(f"model is missing {key}")
    if model["format"] != "GOTANET1":
        raise PackageError("model.format must be GOTANET1")
    hidden, _ = check_model(files["model.bin"], obs_hash, action_hash)
    if model["inputs"] != OBS_SIZE:
        raise PackageError(f"model.inputs must be {OBS_SIZE}")
    if model["hidden"] != hidden:
        raise PackageError("model.hidden does not match model.bin")
    if model["heads"] != HEADS:
        raise PackageError("model.heads must be [8,25,49,4,6]")
    if "goal" in m:
        _keys(m["goal"], {"red", "blue"}, "goal")
        if set(m["goal"]) != {"red", "blue"}:
            raise PackageError("goal needs red and blue")
        _goal(m["goal"]["red"], "goal.red")
        _goal(m["goal"]["blue"], "goal.blue")
    if "decoder" in m:
        d = m["decoder"]
        _keys(d, {"mode", "temperature"}, "decoder")
        mode = d.get("mode", "argmax")
        if mode == "argmax":
            if "temperature" in d:
                raise PackageError("decoder.temperature needs mode sample")
        elif mode == "sample":
            t = d.get("temperature", 1.0)
            if not isinstance(t, (int, float)) or not 0.01 <= t <= 10:
                raise PackageError("decoder.temperature must be 0.01..10")
        else:
            raise PackageError("decoder.mode must be argmax or sample")
    return m


def encode_model(weights, hidden, obs_hash=OBS_HASH, action_hash=ACTION_HASH, inputs=OBS_SIZE, heads=HEADS):
    """GOTANET1 bytes from a flat float32 list: W_enc[H][I], W_rec[3H][H], W_dec[O][H]."""
    outputs = sum(heads)
    expected = inputs * hidden + 3 * hidden * hidden + outputs * hidden
    assert len(weights) == expected
    return (MAGIC + struct.pack("<6I", 1, inputs, hidden, outputs, len(heads), expected) +
            obs_hash.encode() + action_hash.encode() + struct.pack(f"<{len(heads)}I", *heads) +
            struct.pack(f"<{expected}f", *weights))


def build(policy: bytes, model: bytes, period=4, decoder=None, goal=None) -> bytes:
    hidden = struct.unpack_from("<I", model, 16)[0]
    m = {"schema": SCHEMA, "observation_contract": OBS_HASH, "action_contract": ACTION_HASH,
         "decision_period": period,
         "files": {"policy.bas": hashlib.sha256(policy).hexdigest(), "model.bin": hashlib.sha256(model).hexdigest()},
         "model": {"format": "GOTANET1", "inputs": OBS_SIZE, "hidden": hidden, "heads": HEADS}}
    if goal:
        m["goal"] = goal
    if decoder:
        m["decoder"] = decoder
    out = io.BytesIO()
    with zipfile.ZipFile(out, "w", zipfile.ZIP_DEFLATED) as z:
        z.writestr("manifest.json", json.dumps(m, indent=1))
        z.writestr("policy.bas", policy)
        z.writestr("model.bin", model)
    data = out.getvalue()
    validate(data)
    return data


def main():
    ap = argparse.ArgumentParser()
    sub = ap.add_subparsers(dest="cmd", required=True)
    v = sub.add_parser("validate")
    v.add_argument("package")
    v.add_argument("--obs-hash", default=OBS_HASH)
    v.add_argument("--action-hash", default=ACTION_HASH)
    b = sub.add_parser("build")
    b.add_argument("out")
    b.add_argument("--policy", required=True)
    b.add_argument("--model", required=True)
    b.add_argument("--period", type=int, default=4)
    b.add_argument("--decoder", choices=["argmax", "sample"])
    b.add_argument("--temperature", type=float)
    b.add_argument("--goal-red", type=float, nargs=16)
    b.add_argument("--goal-blue", type=float, nargs=16)
    a = ap.parse_args()
    if a.cmd == "validate":
        data = open(a.package, "rb").read()
        if not is_package(data):
            print("plain BASIC submission (not a package): unchanged")
            return
        try:
            m = validate(data, a.obs_hash, a.action_hash)
        except PackageError as e:
            print(f"REJECTED: {e}")
            sys.exit(1)
        print(f"OK {SCHEMA} hidden={m['model']['hidden']} period={m['decision_period']}")
    else:
        decoder = None
        if a.decoder:
            decoder = {"mode": a.decoder}
            if a.temperature is not None:
                decoder["temperature"] = a.temperature
        goal = {"red": a.goal_red, "blue": a.goal_blue} if a.goal_red and a.goal_blue else None
        data = build(open(a.policy, "rb").read(), open(a.model, "rb").read(), a.period, decoder, goal)
        open(a.out, "wb").write(data)
        print(f"wrote {a.out} ({len(data)} bytes)")


if __name__ == "__main__":
    main()
